sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,388 @@
+                async (s, a, c) => await ReadMapPairAsync(s, parseItemAsync, a, c).ConfigureAwait(false);
+            #pragma warning restore AZC0108 // Incorrect 'async' parameter value.
+            #pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            IEnumerable<KeyValuePair<string, T>> entries =
+                await ReadArrayAsync(stream, parsePair, async, cancellationToken).ConfigureAwait(false);
+            return entries.ToDictionary();
+        }
+
+        private static async Task<IEnumerable<T>> ReadArrayAsync<T>(
+            Stream stream,
+            Func<Stream, bool, CancellationToken, Task<T>> parseItemAsync,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            // TODO: This is unpleasant, but I don't want to switch everything to IAsyncEnumerable for every array
+            List<T> items = new List<T>();
+            for (long length = await ReadLongAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                 length != 0;
+                 length = await ReadLongAsync(stream, async, cancellationToken).ConfigureAwait(false))
+            {
+                // Ignore block sizes because we're not skipping anything
+                if (length < 0)
+                {
+                    await ReadLongAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                    length = -length;
+                }
+                while (length-- > 0)
+                {
+                    #pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+                    T item = await parseItemAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                    #pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+                    items.Add(item);
+                };
+            }
+            return items;
+        }
+
+        internal static List<T> Map<T>(this JsonElement array, Func<JsonElement, T> selector)
+        {
+            var values = new List<T>();
+            foreach (JsonElement element in array.EnumerateArray())
+            {
+                values.Add(selector(element));
+            }
+            return values;
+        }
+
+        internal static Dictionary<string, T> ToDictionary<T>(this IEnumerable<KeyValuePair<string, T>> values)
+        {
+            Dictionary<string, T> dict = new Dictionary<string, T>();
+            foreach (KeyValuePair<string, T> pair in values)
+            {
+                dict[pair.Key] = pair.Value;
+            }
+            return dict;
+        }
+    }
+
+    internal abstract class AvroType
+    {
+        public abstract Task<object> ReadAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken);
+
+        public static AvroType FromSchema(JsonElement schema)
```

Comment:
this may be neater if you split it further into FromStringSchema, FromArraySchema, FromObjectSchema

Reply:
Do you think it would be helpful to add constants for each type?

Reply:
Added constants for each type.

Reply:
Split into FromStringSchema, FromArraySchema, FromObjectSchema.